src/utils/dimension-quantity.ts

```diff
@@ -0,0 +1,316 @@
+}
+
+interface QuantityConstructor<T extends Quantity<T>> {
+  new(magnitude: BigNumber): T;
+}
+
+interface VerifiableQuantity<T extends Quantity<T>> extends QuantityConstructor<T> {
+  ZERO: T;
+  isValidMagnitude?(magnitude: BigNumber): undefined | Error; // TODO doc
+}
+
+// TODO the name UnverifiedQuantity shows up when a user fails to expect(); it could probably be named something more instructive, like NeedToExpectQuantity
+// `Type 'UnverifiedQuantity' is not assignable to type 'Tokens'. Property 'tokens' is missing in type 'UnverifiedQuantity'.`
+class UnverifiedQuantity extends Quantity<UnverifiedQuantity> {
+  constructor(magnitude: BigNumber, dimension: DimensionVector) {
+    class Copy extends UnverifiedQuantity {
+      constructor(magnitude: BigNumber) {
+        super(magnitude, dimension);
+      }
+    }
+    super(Copy, magnitude, dimension);
+  }
+  /*
+  TODO an expect failure currently looks like
+  ```
+    Error: expect failed: dimensions not equal, expected=class Tokens extends Quantity {
+    constructor(magnitude) {
+        super(Tokens, magnitude, TokensUnitDimension);
+    }
+  }, actual=[object Object]
+    at UnverifiedQuantity.expect (/Users/ryanandlyndsey/projects/augur-node/src/utils/dimension-quantity.ts:153:13)
+    at Object.<anonymous> (/Users/ryanandlyndsey/projects/augur-node/src/blockchain/log-processors/profit-loss/update-profit-loss.ts:176:73)
+    at Generator.next (<anonymous>)
+    at fulfilled (/Users/ryanandlyndsey/projects/augur-node/src/blockchain/log-processors/profit-loss/update-profit-loss.ts:4:58)
+    at <anonymous>
+  ```
+    instead it should look something like
+      "expect failed: dimensions not equal, expected Price and got Shares"
+      "expect failed: dimensions not equal, expected Price and got { tokens: 1, shares: -5 }"
+  */
+  public expect<T extends Quantity<T>>(vq: VerifiableQuantity<T>): T {
+    if (!isEqual(vq.ZERO.dimension, this.dimension)) {
+      throw new Error(`expect failed: dimensions not equal, expected=${vq}, actual=${this}`);
+    }
+    if (vq.isValidMagnitude !== undefined) {
+      const maybeErr = vq.isValidMagnitude(this.magnitude);
+      if (maybeErr !== undefined) {
+        // TODO wrap maybeErr with dimension context
+        throw maybeErr;
+      }
+    }
+    return new vq(this.magnitude);
+  }
+}
+
+// ************************************************************************
+// **** specific units below here; a lot of this could be generated code
+
+export function scalar(magnitude: number | BigNumber): Scalar {
+  if (typeof magnitude === "number") {
+    return new Scalar(new BigNumber(magnitude));
+  }
+  return new Scalar(magnitude);
+}
+const ScalarUnitDimension: DimensionVector = Object.freeze({
+  tokens: 0,
```

Comment:
Why not the constant ZERO here? Just curious.

Reply:
`ZERO` is a BigNumber. Opted to use regular javascript `number` for the dimension exponents. During implementation was experimenting with including exponent information some types, using typescript "literal types" which require numbers. Am currently satisfied to use regular numbers for exponents; they exception/debug print easier and require fewer allocations, and there's no risk of overflow or need for fancy operations for exponents.